src/cargo/core/compiler/fingerprint/dirty_reason.rs

```diff
@@ -0,0 +1,508 @@
+                s.dirty_because(unit, "the environment variables changed")?;
+
+                Ok(())
+            }
+            DirtyReason::EnvVarChanged { name, .. } => {
+                s.dirty_because(unit, format!("the env variable {name} changed"))?;
+
+                Ok(())
+            }
+            DirtyReason::LocalFingerprintTypeChanged { .. } => {
+                s.dirty_because(unit, "the local fingerprint type changed")?;
+
+                Ok(())
+            }
+            DirtyReason::NumberOfDependenciesChanged { deps } => {
+                Self::present_dependency_diffs(s, unit, deps)
+            }
+            DirtyReason::UnitDependencyNameChanged { deps, .. } => {
+                Self::present_dependency_diffs(s, unit, deps)
+            }
+            DirtyReason::UnitDependencyInfoChanged { deps, .. } => {
+                Self::present_dependency_diffs(s, unit, deps)
+            }
+            DirtyReason::FsStatusOutdated(FsStatusOutdated(status)) => match status {
+                FsStatus::Stale => s.dirty_because(unit, "stale, unknown reason"),
+                FsStatus::StaleItem { item } => match item {
+                    StaleItem::MissingFile(missing_file) => {
+                        let file = missing_file.strip_prefix(root).unwrap_or(&missing_file);
+                        s.dirty_because(unit, format!("the file `{}` is missing", file.display()))
+                    }
+                    StaleItem::ChangedFile {
+                        stale,
+                        stale_mtime,
+                        reference_mtime,
+                        ..
+                    } => {
+                        let file = stale.strip_prefix(root).unwrap_or(&stale);
+                        s.dirty_because(
+                            unit,
+                            format!(
+                                "the file `{}` has changed ({})",
+                                file.display(),
+                                Self::after(*reference_mtime, *stale_mtime, "last build"),
+                            ),
+                        )
+                    }
+                    StaleItem::ChangedEnv { var, .. } => {
+                        s.dirty_because(unit, format!("the environment variable {var} changed"))?;
+                        Ok(())
+                    }
+                },
+                FsStatus::StaleDependency {
+                    name,
+                    dep_mtime,
+                    max_mtime,
+                    ..
+                } => {
+                    let message = match dep_mtime {
+                        Some(dep_mtime) => {
+                            format!(
+                                "the dependency {name} was rebuilt ({})",
+                                Self::after(*max_mtime, *dep_mtime, "last build"),
+                            )
+                        }
+                        None => {
```

Comment:
This branch of `FsStatus::StaleDependency` doesn't use field other than `name`. Maybe we could have the other variant `StaleDepFingerprint(dep_name)` instead.